src/Stratis.Bitcoin/Base/TipsManager.cs

```diff
@@ -0,0 +1,205 @@
+
+        /// <summary>Highest tip commited between all registered components.</summary>
+        private ChainedHeader lastCommonTip;
+
+        /// <summary>Protects all access to <see cref="tipsByProvider"/> and write access to <see cref="lastCommonTip"/>.</summary>
+        private readonly object lockObject;
+
+        /// <summary>Triggered when <see cref="lastCommonTip"/> is updated.</summary>
+        private readonly AsyncManualResetEvent newCommonTipSetEvent;
+
+        private Task commonTipPersistingTask;
+
+        private readonly CancellationTokenSource cancellation;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        public TipsManager(IKeyValueRepository keyValueRepo, ILoggerFactory loggerFactory)
+        {
+            this.keyValueRepo = keyValueRepo;
+            this.tipsByProvider = new Dictionary<object, ChainedHeader>();
+            this.lockObject = new object();
+            this.newCommonTipSetEvent = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <inheritdoc />
+        public void Initialize(ChainedHeader highestHeader)
+        {
+            if (this.commonTipPersistingTask != null)
+                throw new Exception("Already initialized.");
+
+            var commonTipHashHeight = this.keyValueRepo.LoadValue<HashHeightPair>(commonTipKey);
+
+            if (commonTipHashHeight != null)
+                this.lastCommonTip = highestHeader.FindAncestorOrSelf(commonTipHashHeight.Hash, commonTipHashHeight.Height);
+            else
+                // Genesis.
+                this.lastCommonTip = highestHeader.GetAncestor(0);
+
+            this.logger.LogDebug("Tips manager initialized at '{0}'.", this.lastCommonTip);
+
+            this.commonTipPersistingTask = this.PersistCommonTipContinuouslyAsync();
+        }
+
+        /// <summary>Continuously persists <see cref="lastCommonTip"/> to hard drive.</summary>
+        private async Task PersistCommonTipContinuouslyAsync()
+        {
+            while (!this.cancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    await this.newCommonTipSetEvent.WaitAsync(this.cancellation.Token).ConfigureAwait(false);
+                    this.newCommonTipSetEvent.Reset();
+                }
+                catch (OperationCanceledException)
+                {
+                    return;
+                }
+
+                ChainedHeader tipToSave = this.lastCommonTip;
+
+                var hashHeight = new HashHeightPair(tipToSave);
+                this.keyValueRepo.SaveValue(commonTipKey, hashHeight);
+
+                this.logger.LogDebug("Saved common tip '{0}'.", tipToSave);
+            }
+        }
+
+        /// <inheritdoc />
+        public void RegisterTipProvider(object provider)
```

Comment:
Feels a bit wrong to pass an object here.

Reply:
Why even register any component? registration could happen when the component calls set tip.

Reply:
>Feels a bit wrong to pass an object here.

All we need here is ability to compare references so we don't need anything more than just an `object`

>Why even register any component? registration could happen when the component calls set tip.

When we start the node and first component sets the tip we will think that it's the common tip, when other components might have different tips. And we will save wrong value to db. If node shuts down after that we won't be able to recover next time we start the node.

Reply:
String probably won't work since we might accidentally have collisions. 
Type will work as long as we know for sure we won't have 2 instances of the same type having their own tips. 

But still just comparing the reference using object provides the most flexible solution.

Reply:
I don't get why not using an interface.
The interface has to be tought as a contract, and if we have an interface called, let say, `IManagedTip`, you know that any component that uses that interface has the tip managed by the TipManager

This way we can get rid of the `RegisterTipProvider(object)` and just have RegisterTipProvider(IManagedTip).
This doesn't mean the solution is more complex, on the contrary, and you clearly declare your intention on components, because the interface IManagedTip declare the component has a tip, and it's managed.

With this change, you can then change your dictionary to be of type `Dictionary<IManagedTip, ChainedHeader>`.
Much more clean without having an object and the only need is to add the interface to components using the tipmanager, but you already need a change to that classes anyway to register anyway.

As a side node, consider that having an interface could even allow us to cycle all registered service that implement this interface and autoregister them

ah, of course even CommitTipPersisted should reference the interface rather then the object, but those are very trivial changes

Reply:
Such an interface will be empty, won't it? 

It feels weird to me to have an empty interface there just to avoid using bare object.
But I can do it. 

@dangershony what do you think?

Reply:
I'm not sure it will stay empty, e.g. having a GetCurrentTip could be useful.
I'm not sure about CommitTipPersisted neither because it causes lot of locks during sync, not sure we want it,, but I'll look at it later

Reply:
>e.g. having a GetCurrentTip could be useful.

Some components have HashHeight pair as their tip representation, some have ChainedHeader. 

>I'm not sure about CommitTipPersisted neither because it causes lot of locks during sync

Not really, in IBD we always save stuff in batches. Plus the code inside a lock is pretty lightweighted

Reply:
I'm definitely in favor of the empty interface and not passing `object`. Its like you are asking for trouble by doing that ;)

Reply:
Ok, fixed 🙂